source/models/Activator.py

```diff
@@ -0,0 +1,47 @@
+from abc import ABC, abstractmethod
+
+from util.InvalidStateException import InvalidStateException
+
+
+class Activator(ABC):
+    """
+    An abstract implementation of an activator.
+    Concrete activators should only implement the abstract method perform.
+    """
+
+    def __init__(self, database, device_id, activator_id):
+        self._database = database
+        self._device_id = device_id
+        self._activator_id = activator_id
+
+    @property
+    def identifier(self):
+        return self._activator_id
+
+    @property
+    def rank(self):
```

Comment:
some methods in this project could still benefit from docstrings. It is not, for example, immediately clear why activators have a 'rank'.